### ISSUE TITLE
Add the ability to pass a uv_handle_t to a new event loop

### ISSUE THREAD

**Issue by kellabyte** [libuv/libuv#390]
Copying this issue that is over a year old from the old Joyent repository:
https://github.com/joyent/libuv/issues/1301

It would be really great to avoid having to implement the complexity of https://github.com/joyent/libuv/blob/master/test/benchmark-multi-accept.c and avoid sending over IPC to load balance on other event loops. I find this code hard to follow and it's the part I struggle with the most out of all my libuv calling code.

Potentially latency can be improved if this isn't done over IPC?

One issue that was raised is how this is more difficult on Windows because of IO completion ports (IOCP) and thread ownership. I spoke to Microsoft and I received the following response about how to solve this on Windows:

> The NtSetInformationFile function with FileReplaceCompletionInformation parameter accomplishes this. You can swap in a new IOCP, or just clear out the existing IOCP (by passing in NULL for FILE_COMPLETION_INFORMATION.Port). There must not be any pending I/O requests on the socket/handle for this to succeed.

**Comment by saghul** [comment]
I did (very briefly) look into this a while ago, sorry for not coming back to report :-S

> One issue that was raised is how this is more difficult on Windows because of IO completion ports (IOCP) and thread ownership. I spoke to Microsoft and I received the following response about how to solve this on Windows

Sadly this is seems to be supported only on Windows >= 8.1:

~~~~
A FILE_COMPLETION_INFORMATION structure to change or remove the completion port associated with a file handle. This value is available starting with Windows 8.1.
~~~~

https://msdn.microsoft.com/en-us/library/windows/hardware/ff728840%28v=vs.85%29.aspx

**Comment by benaadams** [comment]
That's also >= Windows Server 2012 R2?

**Comment by addaleax** [comment]
I’d love to look into this. What I imagine is a pair of functions, one detaching an (inactive?) handle from one event loop, one attaching it to another one.

What would be the largest roadblocks for that? I assume that it’s not feasible for all types of handles, but for most/a lot of them?

**Comment by bnoordhuis** [comment]
Yes, it's applicable to all file descriptor-backed handle types.

<sup>There's no reason you couldn't implement logic to handle `uv_async_t` or `uv_timer_t` handles as well but that's of limited value.  People mainly want to send TCP and UDP handles to another thread.</sup>

It's pretty easy on Unices.  The big obstacle is Windows - moving handles from one completion port to another is complicated.

For locking and waking up the receiving event loop, you can piggyback on `loop->wq_mutex` and `loop->wq_async`.

**Comment by addaleax** [comment]
> For locking and waking up the receiving event loop, you can piggyback on loop->wq_mutex and loop->wq_async.

Just so we’re on the same page, do we need that if we make it a requirement that the import/export functions are run on the same threads as their corresponding event loops? Or is that an unreasonable requirement?

**Comment by bnoordhuis** [comment]
You wouldn't need it in that case, but users would then need to devise their own transfer mechanism.

But you raise a good point.  We should probably start out simple and lift the restriction in the future if there is demand for it.

**Comment by vtjnash** [comment]
> Sadly this is seems to be supported only on Windows >= 8.1:

I guess this is now legal to implement :)
